### PR TITLE
divmodw && gaid

### PR DESCRIPTION
**Pull request by shiqizng**
Adding support for opcode divmodw and gaid(s)

### CODE REVIEW

File: TealangLexer.l4 [hunk 1]
```diff
@@ -137,6 +137,8 @@ ARGS        : 'args' ;
 ACCOUNTS    : 'accounts' ;
 APPS        : 'apps' ;
 ASSETS      : 'assets' ;
+EXTRAPROGRAMPAGES    : 'ExtraProgramPages';
+
```
Comment (algorandskiy) [inline TealangLexer.l4:141]: please move it back to line 208 where all other txn fields are defined

File: TealangParser.g4 [hunk 2]
```diff
@@ -64,13 +64,15 @@ termination
 decl
     :   LET IDENT EQ expr                          # DeclareVar
     |   LET IDENT COMMA IDENT EQ tupleExpr         # DeclareVarTupleExpr
+    |   LET IDENT COMMA IDENT COMMA IDENT COMMA IDENT EQ tupleExpr # DeclareVarTupleExpr
```
Comment (algorandskiy) [inline TealangParser.g4:67]: antlr should complain about duplicate names like `DeclareVarTupleExpr`

File: compiler/ast.go [hunk 3]
```diff
@@ -263,6 +264,10 @@ type TreeNode struct {
 	childrenNodes []TreeNodeIf
 }
 
+type forNode struct {
+	*TreeNode
+}
+
```
Comment (algorandskiy) [inline compiler/ast.go:270]: not needed

File: compiler/ast_test.go [hunk 4]
```diff
@@ -48,7 +48,7 @@ function logic() {
 	}
 
 	let y= 0;
-	for 1 {
+	for 1 {
```
Comment (algorandskiy) [inline compiler/ast_test.go:51]: not needed change

File: compiler/codegen.go [hunk 5]
```diff
@@ -157,6 +157,13 @@ func (n *varDeclTupleNode) Codegen(ostream io.Writer) {
 	fmt.Fprintf(ostream, "store %d\n", info.address)
 	info, _ = n.ctx.lookup(n.high)
 	fmt.Fprintf(ostream, "store %d\n", info.address)
+	if n.nodeName == "divmodw" {
+		info, _ = n.ctx.lookup(n.rlow)
+		fmt.Fprintf(ostream, "store %d\n", info.address)
+		info, _ = n.ctx.lookup(n.rhigh)
+		fmt.Fprintf(ostream, "store %d\n", info.address)
+	}
+
```
Comment (algorandskiy) [inline compiler/codegen.go:166]: it looks like would be better to define `varDeclQuadrupleNode` and do not have `divmodw` mention here

**Review by algorandskiy — COMMENTED** [review]
Looks good but I think it might be done better with introducing two new rules `DeclareVarQuadrupleExpr` and `AssignQuadruple`.

Please also add an e2e test [similar to expw](https://github.com/pzbitskiy/tealang/blob/master/test/tuple_test.go#L50) test

**Comment by shiqizng** [comment]
e2e testing is failing with error "[{errorType:3 start:112 end:117 line:6 column:1 msg:incompatible types: 'invalid' vs 'uint64' in expr 'ident rlow == 0' token:assert filename: excerpt:[    assert(rlow == 0) ----^-----]}]"

**Review by pzbitskiy — COMMENTED** [review]
`divmodw` looks good.
`gaid` requires some additional work:
 - it should be a built-in function rather object.
 - no need to introduce both `gaid` and `gaids` to tealang - the compiler can be smart enough and use `gaid` or `gaids` depend on the arg type. See remapper and substring example in `builtin.go` (requires rebasing). In short, `substring` handler analyzes arguments type and select imm TEAL opcode if both of them are constants. Another example is gtxn vs gtxns.

File: compiler/ast.go [hunk 6]
```diff
@@ -847,6 +912,35 @@ func (n *funCallNode) getTypeTuple() (exprType, exprType, error) {
...
+	}
+	tpl, err = opTypeFromSpec(n.name, 1)
+	if err != nil {
+		return tph, tpl, rtpl, rtph, err
+	}
+	rtpl, err = opTypeFromSpec(n.name, 2)
+	if err != nil {
+		return tph, tpl, rtpl, rtph, err
+	}
+	rtpl, err = opTypeFromSpec(n.name, 3)
```
Comment (pzbitskiy) [inline compiler/ast.go:940]: copy-paste error, should be `rtpl` here and `rtph` above

File: test/tuple_test.go [hunk 7]
```diff
@@ -57,3 +57,17 @@ function logic() {
 }`
 	performTest(t, source)
 }
+
+func TestDivmodw(t *testing.T) {
+	source := `
+function logic() {
+	let qhigh, qlow, rhigh, rlow = divmodw(2, 0, 0, 1)
```
Comment (pzbitskiy) [inline test/tuple_test.go:64]: let's divide 99/5 to check non-zero reminder to ensure correct high-low order

**Comment by shiqizng** [comment]
opened separate PRs

**Pull request closed without merging**